src/qrules/io/_dot.py

```diff
@@ -497,26 +495,49 @@ def _collapse_graphs(
         )
         for g in graphs
     }
+
+    sorting_keys = []
     for transition in graphs:
         topology = transition.topology
         group = transition_groups[topology]
         for state_id, state in transition.states.items():
-            if isinstance(state, State):
-                particle = state.particle
-            else:
-                particle, _ = state
-            group.states[state_id].add(particle)
+            group.states[state_id].add(_strip_properties(state))
+            sorting_keys.append(_gen_sorting_key(state))
     collected_graphs: list[FrozenTransition[tuple[Particle, ...], None]] = []
     for topology in sorted(transition_groups):
         group = transition_groups[topology]
         collected_graphs.append(
             FrozenTransition(
                 topology,
                 states={
-                    i: tuple(sorted(particles, key=lambda p: p.name))
+                    i: tuple(sorted(particles, key=sorting_keys[i]))
                     for i, particles in group.states.items()
                 },
                 interactions=group.interactions,
             )
         )
     return collected_graphs
+
+
+def _strip_properties(state: Any) -> Any:
+    if isinstance(state, State):
+        return state.particle
+    if isinstance(state, str):
+        return state
+    if isinstance(state, list):
+        return tuple(state)
+    if isinstance(state, tuple):
+        return state
+    return state
+
+
+def _gen_sorting_key(obj: Any) -> Callable:
+    if isinstance(obj, State):
+        return lambda part: part.name
+    if isinstance(obj, str):
+        return str.lower
+    if isinstance(obj, list):
+        return _gen_sorting_key(obj[0])
+    if isinstance(obj, tuple):
+        return _gen_sorting_key(obj[0])
+    return operator.__lt__
```

Comment:
Seems to me an overcomplication of the logic. All this function has to do is turn any object (any item in the iterable that is to be sorted) into something that supports comparison (`__lt__()`). See [my suggestion](https://github.com/ComPWA/qrules/pull/274#discussion_r1668674840). That would also remove the need for collecting all these callables in `sorting_keys`.

Reply:
I realized btw that hashable is not the proper return type. There is no good built-in type for something that is 'comparable', but you may use `SupportsRichComparison`, see e.g. [this example](https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-types-defined-in-stubs-but-not-at-runtime).